lib/apiservers/engine/backends/cache/repo_cache.go

```diff
@@ -0,0 +1,416 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/kv"
+	"github.com/vmware/vic/lib/apiservers/portlayer/client"
+	"github.com/vmware/vic/pkg/trace"
+
+	"github.com/docker/distribution/digest"
+	"github.com/docker/docker/reference"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// repoCache is a cache of the docker repository information.
+// This info will help to provide proper tag and digest support
+//
+// The cache will be persisted to disk via the portlayer k/v
+// store and will be restored at system start
+//
+// This code is a heavy leverage of docker's reference store:
+// github.com/docker/docker/reference/store.go
+
+var (
+	rCache  *repoCache
+	repoKey = "repositories"
+)
+
+// Repo provides the set of methods which can operate on a tag store.
+type Repo interface {
+	References(imageID string) []reference.Named
+	ReferencesByName(ref reference.Named) []Association
+	Delete(ref reference.Named, save bool) (bool, error)
+	Get(ref reference.Named) (string, error)
+
+	Save() error
+	GetImageID(layerID string) string
+	Tags(imageID string) []string
+	Digests(imageID string) []string
+	AddReference(ref reference.Named, imageID string, force bool, layerID string, save bool) error
+
+	// Remove will remove from the cache and returns the
+	// stringified Named if successful -- save bool instructs
+	// func to persist to portlayer k/v or not
+	Remove(ref string, save bool) (string, error)
+}
+
+type repoCache struct {
+	// client is needed for k/v store operations
+	client *client.PortLayer
+
+	mu sync.RWMutex
```

Comment:
Since there can be only one `repoCache` (in the form of `rCache`), you may want to consider moving this mutex to the same level as `rCache`.  Then you can avoid the assignment/nil race on startup.